homeassistant/components/zha/core/store.py

```diff
@@ -0,0 +1,146 @@
+"""Data storage helper for ZHA."""
+import logging
+from collections import OrderedDict
+# pylint: disable=W0611
+from typing import MutableMapping  # noqa: F401
+from typing import cast
+
+import attr
+
+from homeassistant.core import callback
+from homeassistant.loader import bind_hass
+from homeassistant.helpers.typing import HomeAssistantType
+
+_LOGGER = logging.getLogger(__name__)
+
+DATA_REGISTRY = 'zha_storage'
+
+STORAGE_KEY = 'zha.storage'
+STORAGE_VERSION = 1
+SAVE_DELAY = 10
+
+
+@attr.s(slots=True, frozen=True)
+class ZhaDeviceEntry:
+    """Zha Device storage Entry."""
+
+    name = attr.ib(type=str, default=None)
+    ieee = attr.ib(type=str, default=None)
+    power_source = attr.ib(type=int, default=None)
+    manufacturer_code = attr.ib(type=int, default=None)
+
+
+class ZhaDeviceStorage:
+    """Class to hold a registry of zha devices."""
+
+    def __init__(self, hass: HomeAssistantType) -> None:
+        """Initialize the zha device storage."""
+        self.hass = hass
+        self.devices = {}  # type: MutableMapping[str, ZhaDeviceEntry]
+        self._store = hass.helpers.storage.Store(STORAGE_VERSION, STORAGE_KEY)
+
+    @callback
+    def async_create(self, device) -> ZhaDeviceEntry:
+        """Create a new ZhaDeviceEntry."""
+        device_entry = ZhaDeviceEntry(
+            name=device.name,
+            ieee=str(device.ieee),
+            power_source=device.power_source,
+            manufacturer_code=device.manufacturer_code
+
+        )
+        self.devices[device_entry.ieee] = device_entry
+
+        return self.async_update(device)
+
+    @callback
+    def async_get_or_create(self, device) -> ZhaDeviceEntry:
+        """Create a new ZhaDeviceEntry."""
+        ieee_str = str(device.ieee)
+        if ieee_str in self.devices:
+            return self.devices[ieee_str]
+        return self.async_create(device)
+
+    @callback
+    def async_create_or_update(self, device) -> ZhaDeviceEntry:
+        """Create or update a ZhaDeviceEntry."""
+        if str(device.ieee) in self.devices:
+            return self.async_update(device)
+        return self.async_create(device)
+
+    async def async_delete(self, ieee: str) -> None:
+        """Delete ZhaDeviceEntry."""
+        del self.devices[ieee]
+        self.async_schedule_save()
+
+    @callback
+    def async_update(self, device) -> ZhaDeviceEntry:
+        """Update name of ZhaDeviceEntry."""
+        ieee_str = str(device.ieee)
+        old = self.devices[ieee_str]
+
+        changes = {}
+
+        if device.power_source != old.power_source:
+            changes['power_source'] = device.power_source
+
+        if device.manufacturer_code != old.manufacturer_code:
+            changes['manufacturer_code'] = device.manufacturer_code
+
+        new = self.devices[ieee_str] = attr.evolve(old, **changes)
```

Comment:
So there are two approaches to dealing with stored entries. What you're doing here is treating all entries as immutable, and so anyone that has fetched an object in the past will now keep the old one. You can also choose to update in-place, so that places have updates right away (but might now know about it).

You pick what you want 👍 Just wanted to make sure you know you've got options.

Reply:
Cool, what does the alternative look like code wise? Got an example handy?

Reply:
@balloob I decided to leave this like this.